src/st/nz0/gaibon.c

```diff
@@ -0,0 +1,708 @@
+#include "nz0.h"
+
+#define SLOGRA self[-8]
+
+typedef enum {
+    GAIBON_INIT,
+    GAIBON_IDLE = 2,
+    GAIBON_FLY_TOWARDS_PLAYER,
+    GAIBON_FLY_SHOOT_FIREBALLS,
+    GAIBON_LANDING_AFTER_SHOOTING,
+    GAIBON_SHOOT_FROM_GROUND,
+    GAIBON_FLY_SHOOT_BIG_FIREBALL = 8,
+    GAIBON_PICKUP_SLOGRA,
+    GAIBON_NEAR_DEATH = 15,
+    GAIBON_DYING,
+    GAIBON_DEBUG = 255,
+} GaibonSteps;
+
+typedef enum {
+    GAIBON_FLY_TOWARDS_PLAYER_BEGIN,
+    GAIBON_FLY_TOWARDS_PLAYER_MOVEMENT,
+    GAIBON_FLY_TOWARDS_PLAYER_END,
+} GaibonFlyTowardsPlayerSubSteps;
+
+typedef enum {
+    GAIBON_FLY_SHOOT_FIREBALLS_BEGIN,
+    GAIBON_FLY_SHOOT_FIREBALLS_MOVING_SHOOTING,
+    GAIBON_FLY_SHOOT_FIREBALLS_END,
+} GaibonFlyShootFirewalls_SubSteps;
+
+typedef enum {
+    GAIBON_LANDING_AFTER_SHOOTING_SETUP,
+    GAIBON_FALLING_WITHOUT_MAP_COLLISION,
+    GAIBON_FALLING_WITH_MAP_COLLISION,
+    GAIBON_FALLING_ON_GROUND,
+} GaibonLandingAfterShootingSubSteps;
+
+typedef enum {
+    GAIBON_SHOOT_FROM_GROUND_FACE_PLAYER,
+    GAIBON_SHOOT_FROM_GROUND_FACE_SETUP,
+    GAIBON_SHOOT_FROM_GROUND_FACE_SHOOTING,
+} GaibonShootFromGroundSubSteps;
+
+typedef enum {
+    GAIBON_FLY_SHOOT_BIG_FIREBALL_SETUP,
+    GAIBON_FLY_SHOOT_BIG_FIREBALL_SHOOTING,
+    GAIBON_FLY_SHOOT_BIG_FIREBALL_END,
+} GaibonFlyShootBigFireballSubSteps;
+
+typedef enum {
+    GAIBON_PICKUP_SLOGRA_SETUP,
+    GAIBON_PICKUP_SLOGRA_MOVING,
+    GAIBON_PICKUP_SLOGRA_ASCENDING,
+    GAIBON_PICKUP_SLOGRA_AIMING,
+    GAIBON_PICKUP_SLOGRA_RELEASE,
+} GaibonPickupSlograSubSteps;
+
+typedef enum {
+    GAIBON_NEAR_DEATH_SETUP,
+    GAIBON_NEAR_DEATH_FLOOR_HIT_WAIT,
+    GAIBON_NEAR_DEATH_FLOOR_LANDING,
+    GAIBON_NEAR_DEATH_TRANSFORM,
+} GaibonNearDeathSubSteps;
+
+typedef enum {
+    GAIBON_DYING_REACT,
+    GAIBON_DYING_TURN_INTO_BONES,
+} GaibonDyingSubSteps;
+
+void EntityGaibon(Entity* self) {
+    Collider collider;
+    Entity* other;
+    s16 angle;
+    s32 speed;
+    s32 speedLimit;
+    s8* hitboxPtr;
+
+    s32 yVar;
+    s32 xVar;
+
+    s32 stupid;
+
+    if ((self->step) && (!self->ext.GS_Props.nearDeath) &&
+        self->hitPoints < g_api.enemyDefs[0xFE].hitPoints / 2) {
```

Comment:
> g_api.enemyDefs[0xFE]

I wonder if there is a better way to reference that `0xFE` to the actual enemy entry. I do not think an enum would suffice here. Let's keep it as it is but explore what we can do later on